src/tickv.md

```diff
@@ -88,61 +88,64 @@ field, then it can change the value of that key-value object.
-    fn generate_key(&self, unhashed_key: [u8], key: K) -> Result<(), (K, buffer,ErrorCode)>;
+pub trait KV {
+    /// Retrieve a value from the store.
+    fn get(&self, key: [u8], value: [u8]) -> Result<(), ([u8], [u8], ErrorCode)>;
 
-    /// Add a K-V object to the store. Error on collision.
-    fn append_key(&self, key: K, value: [u8]) -> Result<(), (K, buffer, ErrorCode)>;
+    /// Insert a key-value object into the store. Overwrite if needed.
+    fn set(&self, key: [u8], value: [u8]) -> Result<(), ([u8], [u8], ErrorCode)>;
 
-    /// Retrieve a value from the store.
-    fn get_value(&self, key: K, value: [u8]) -> Result<(), (K, buffer, ErrorCode)>;
+    /// Insert a key-value object into the store if it doesn't exist.
+    fn add(&self, key: [u8], value: [u8]) -> Result<(), ([u8], [u8], ErrorCode)>;
 
-    /// Mark a K-V object as deleted.
-    fn invalidate_key(&self, key: K) -> Result<(), (K, ErrorCode)>;
+    /// Modify a key-value object into the store if it already exists.
+    fn update(&self, key: [u8], value: [u8]) -> Result<(), ([u8], [u8], ErrorCode)>;
 
-    /// Cleanup the store.
-    fn garbage_collect(&self) -> Result<(), ErrorCode>;
+    /// Remove a key-value object from the store.
+    fn update(&self, key: [u8]) -> Result<(), ([u8], ErrorCode)>;
 }
 ```
 
-(You can find the full definition in `tock/kernel/src/hil/kv_system.rs`.)
+(You can find the full definition in `tock/kernel/src/hil/kv.rs`.)
 
-In terms of TicKV, the `KVSystem` interface only uses the TicKV header. The Tock
-header is only used in the upper level API.
+To enable access control, we layer an additional KV interface on top of the `KV`
+interface.
 
 ```rust
-pub trait KVStore {
-    /// Get key-value object.
-    pub fn get(&self, key: [u8], value: [u8], perms: StoragePermissions) -> Result<(), (buffer, buffer, ErrorCode)>;
+pub trait KVPermissions {
+    /// Retrieve a value from the store.
+    fn get(&self, key: [u8], value: [u8], permissions: Perm) -> Result<(), ([u8], [u8], ErrorCode)>;
+
+    /// Insert a key-value object into the store. Overwrite if needed.
+    fn set(&self, key: [u8], value: [u8], permissions: Perm) -> Result<(), ([u8], [u8], ErrorCode)>;
+
+    /// Insert a key-value object into the store if it doesn't exist.
+    fn add(&self, key: [u8], value: [u8], permissions: Perm) -> Result<(), ([u8], [u8], ErrorCode)>;
 
-    /// Set or update a key-value object.
-    pub fn set(&self, key: [u8], value: [u8], perms: StoragePermissions) -> Result<(), (buffer, buffer, ErrorCode)>;
+    /// Modify a key-value object into the store if it already exists.
+    fn update(&self, key: [u8], value: [u8], permissions: Perm) -> Result<(), ([u8], [u8], ErrorCode)>;
 
-    /// Delete a key-value object.
-    pub fn delete(&self, key: [u8], perms: StoragePermissions) -> Result<(), (buffer, ErrorCode)>;
+    /// Remove a key-value object from the store.
+    fn update(&self, key: [u8], permissions: Perm) -> Result<(), ([u8], ErrorCode)>;
```

Comment:
```suggestion
    fn delete(&self, key: [u8], permissions: Perm) -> Result<(), ([u8], ErrorCode)>;
```
?

There are two `update`s in a row right now

Reply:
Or if it's really that "update" now handles both modify and delete operations, make make that a bit clearer somehow?